Fixed black screen on iOS launch

nog1potje opened the pull request:
When launching an iOS game made in defold, a brief black screen could be observed in the transition from the launch screen to the game. This fix covers up that black screen with the contents of the launch screen.

Fixes #6142

- WARNING: as of now this fix is incomplete, as it only works if the user uses a file called LaunchScreen. Please help provide guidance on how to access user set variables.

### Technical changes
- Load the launch screen storyboard and put it in front of the main view as soon as didFinishLaunchingWithOptions is called by the OS
- Remove this cover when the engine is initialized

nog1potje committed:
Fix for 6142 Black screen on iOS launch

- load the launch screen storyboard and put it in front of the main view
- remove this cover when the engine is initialized

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/defold/defold?pullRequest=8995) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/defold/defold?pullRequest=8995) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/defold/defold?pullRequest=8995) it.</sub>

britzl commented on external/glfw/lib/ios/app/AppDelegate.m:
```diff
@@ -182,6 +194,16 @@ - (void)appUpdate
         ShutdownEngine(true);
     }
+
+    // Cleanup the placeholder laucnh screen view once the engine is initialized
```
laucnh -> launch

nog1potje commented:
Does this class have access to dmConfigFile? There's no other usage within the glfw library that I could use as an example. Also, how do I compile the glfw library to test my code? The makefile does not seem to contain a configuration for iOS. I've sent a friend request through discord if that's ok.

britzl commented:
> Does this class have access to dmConfigFile? There's no other usage within the glfw library that I could use as an example. Also, how do I compile the glfw library to test my code? The makefile does not seem to contain a configuration for iOS. I've sent a friend request through discord if that's ok.

Ah, I did not notice that t he fix was in the glfw layer. Ok, we need some additional input then. @Jhonnyg can you assist?

JCash commented on external/glfw/lib/ios/app/AppDelegate.m:
```diff
@@ -84,6 +84,18 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     window.rootViewController = [[[ViewController alloc] init] autorelease];
+
+    // Load the LaunchScreen storyboard
+    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"LaunchScreen" bundle:nil];
```
Agreed, we need to look into the plist properties after the `launch-screen` property before accepting this PR.

JCash reviewed (changes requested):
Overall good, but needs a solution for getting the launch screen name

nog1potje commented:
I figured out how to access the launch screen through iOS functionality, no need to access the defold settings. Commit incoming

nog1potje committed:
Updated fix using launch screen from info.plist

additionally:
- now use existing reference to the window (g_ApplicationWindow) for cleanup
- Cleanup now also checks if a UIView is actually found, which might not be the case if there's no launch screen storyboard

nog1potje commented:
Done, the launch screen is now retrieved from the info.plist dictionary, and will be displayed if found. This should ensure it's always the same visuals as those displayed during load.

britzl commented:
The glfw packages are precompiled using the `build_external ` build command and then copied to `packages`. This PR should preferably also include the pre-built glfw libraries.

https://github.com/defold/defold/tree/dev/external#external

nog1potje commented:
Sounds good, I'll get on it

britzl commented:
> Sounds good, I'll get on it

Cool, let us know if you need help with the setup.

nog1potje committed:
Update glfx package for iOS

nog1potje commented:
glfw successfully compiled and I've committed the file, but my local editor instance won't bundle iOS apps, not sure why. I'm getting com.defold.extender.ExtenderException: The given sdk does not exist: 393881b328f0763b0ba1197764a322a6771869de However, this happens for me on the base dev branch as well. Can someone help test it?

JCash commented:
I think the issue is that the editor tries to build with an archived release, but since you're actually using the local build, it _won't_ be an archived release.

What are your build steps for the editor?
It should be like:

First, for the engine:
* build iOS
* build Bob

And then

```
cd editor
lein init
lein run
```

nog1potje commented:
Thanks for the approval! I haven't been able to put any time in this week unfortunately. Hopefully next week. Also, why the conflict? Are there any other changes to the ios glfw library?

britzl commented:
@nog1potje the comment in the initial PR about _"WARNING: as of now this fix is incomplete, as it only works if the user uses a file called LaunchScreen. Please help provide guidance on how to access user set variables."_ is no longer accurate right?

nog1potje commented:
Yes, that warning can now be ignored

britzl committed:
Merge branch 'dev' into pr/8995

britzl committed:
Update glfw-2.7.1-arm64-ios.tar.gz
